woocommerce/woocommerce-gateway-stripe
# Use metadata stored on the intent to fetch an order for webhook processing

## Changes proposed in this Pull Request:

This PR updates the webhook handler to enable it to fetch the relevant order from `payment_intent` metadata.

This has two impacts:

1. If we fail to record the intent ID on the order after we've sent the payment intent request to Stripe, we can use this mechanism to process the response from Stripe. 
2. It improves the performance of webhook processing because we no longer run a database query to locate an order by the intent ID.

We've seen at least 1 large store where we fire off the payment intent request to Stripe, however, we never process the response. This means we don't record the intent ID on the order and therefore cannot process the webhook when we receive it. This PR adds a way for us to fetch the order from data stored in the intent's metadata. Internal link to this case: pcJe93-4NE-p2#comment-4929

cc @geektzu

## Testing instructions

1. This should have no impact on payments. 
2. You should notice two changes:
   - Payments in the Stripe Dashboard will now record a "signature" in metadata. This value will contain the order ID followed by a hash.
   - The webhook processing will now use the order ID in this signature to validate and process the webhook. 

<p align="center">
<img width="265" alt="Screenshot 2024-10-31 at 3 40 18 pm" src="https://github.com/user-attachments/assets/be60ad50-6a3d-4b3b-bf3e-fc1391212515">
</br>
<sup>The new <strong>signature</strong> stored in payment intent metadata.</sup>
</p>

Because this changes a core part of the way we process webhook payments, please scrutinize this change to make sure we root out any potential gotchas. 

---


**Post merge**
